pennylane/math/quantum.py

```diff
@@ -164,6 +168,173 @@ def marginal_prob(prob, axis):
+
+    if check_state:
+        # Check format
+        if (
+            len(density_matrix.shape) != 2
+            or density_matrix.shape[0] != density_matrix.shape[1]
+            or not np.log2(shape).is_integer()
+        ):
+            raise ValueError("Density matrix must be of shape (2**N, 2**N).")
+        # Check trace
+        trace = np.trace(density_matrix)
+        if not is_abstract(trace):
+            if not allclose(trace, 1.0, atol=1e-10):
+                raise ValueError("The trace of the density matrix should be one.")
+            # Check if the matrix is hermitian
+            conj_trans = np.transpose(np.conj(density_matrix))
+            if not allclose(density_matrix, conj_trans):
+                raise ValueError("The matrix is not hermitian.")
+
+    consecutive_indices = list(range(0, num_indices))
+
+    # Return the full density matrix if all the wires are given
+    if tuple(indices) == tuple(consecutive_indices):
+        return density_matrix
+
+    traced_wires = [x for x in consecutive_indices if x not in indices]
+    density_matrix = partial_trace(density_matrix, traced_wires)
+    return density_matrix
+
+
+def partial_trace(density_matrix, indices, c_dtype="complex64"):
+    """Compute the reduced density matrix by tracing out the provided indices.
+
+    Args:
+        density_matrix (tensor_like): 2D density matrix tensor. This tensor should be of size ``(2**N, 2**N)`` for some
+            integer number of wires ``N``.
+        indices (list(int)): List of indices to be traced.
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        tensor_like: (reduced) Density matrix of size ``(2**len(wires), 2**len(wires))``
+
+    **Example**
+
+    >>> x = np.array([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
+    >>> partial_trace(x, indices=[0])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+
+
+    >>> x = tf.Variable([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=tf.complex128)
+    >>> partial_trace(x, indices=[1])
+    tf.Tensor(
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
+
+    >>> y = [[0.5, 0, 0.5, 0], [0, 0, 0, 0], [0.5, 0, 0.5, 0], [0, 0, 0, 0]]
+    >>> partial_trace(y, indices=[1])
+    [[0.5+0.j 0.5+0.j]
+     [0.5+0.j 0.5+0.j]]
+
+    >>> partial_trace(y, indices=[0])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+    """
+    # Dimension and reshape
+    shape = density_matrix.shape[0]
+    num_indices = int(np.log2(shape))
+    rho_dim = 2 * num_indices
+
+    density_matrix = np.reshape(density_matrix, [2] * 2 * num_indices)
+
+    # Kraus operator for partial tracee
+    kraus = cast(np.eye(2), dtype=c_dtype)
+    kraus = np.reshape(kraus, (2, 1, 2))
+    kraus_dagger = np.asarray([np.conj(np.transpose(k)) for k in kraus])
+
+    kraus = convert_like(kraus, density_matrix)
+    kraus_dagger = convert_like(kraus_dagger, density_matrix)
+
+    # For loop over wires
+    for target_wire in indices:

```

Comment:
The logic here seems to be very similar to a part of the `_apply_channel` of the mixed device. Could `partial_trace` be used there?